src/lib/tls/tls13/tls_client_impl_13.cpp

```diff
@@ -327,7 +327,15 @@ void Client_Impl_13::handle(const Server_Hello_13& sh) {
    } else {
       m_resumed_session.reset();  // might have been set if we attempted a resumption
       m_cipher_state = Cipher_State::init_with_server_hello(
-         m_side, std::move(shared_secret), cipher.value(), m_transcript_hash.current());
+         m_side,
+         std::move(shared_secret),
+         cipher.value(),
+         m_transcript_hash.current(),
+         [this](const char* label, const secure_vector<uint8_t>& secret) {
+            if(policy().allow_ssl_key_log_file()) {
+               callbacks().tls_ssl_key_log_data(label, m_handshake_state.client_hello().random(), secret);
+            }
+         });
```

Comment:
Frankly, I don't mind the extra parameters. It establishes an explicit dependency and therefore communicates that the callbacks and policy are needed in the implementation. There are many examples where several of those "bookkeeping objects" are passed into constructors and methods of objects for such reasons.

I'd definitely prefer passing the `Policy&` and/or `Callbacks&` explicitly into the methods that require it, rather than into the `Cipher_State` constructor, however.

Reply:
oh, like this
```
advance_with_server_hello(const Ciphersuite& cipher,
                                                 secure_vector<uint8_t>&& shared_secret,
                                                 const Transcript_Hash& transcript_hash,
                                                 const Policy& policy,
                                                 const Callbacks& callbacks 
)
```
if you are agree on this, ok.